File: android/app/src/main/java/tw/stsa/memberapp/feature/card/MembershipCodeStore.kt
package tw.stsa.memberapp.feature.card

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import tw.stsa.memberapp.R
import tw.stsa.memberapp.auth.AuthManager
import java.time.Instant
import kotlin.time.Duration.Companion.seconds

/**
 * Fetches and keeps fresh the code behind the member card's QR.
 *
 * The server mints a new code per request and stores it in Redis for 300
 * seconds (`CODE_TTL` in MembershipAPI). The web card re-fetches every 250s to
 * stay inside that window; this does the same, but only while the card is on
 * screen — there is no reason to hold a live code in the user's pocket.
 */
class MembershipCodeStore(
    private val context: Context,
    private val scope: CoroutineScope,
) {
    var payload by mutableStateOf<String?>(null)
        private set
    var issuedAt by mutableStateOf<Instant?>(null)
        private set
    var isRefreshing by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    private var refreshJob: Job? = null

    val expiresAt: Instant? get() = issuedAt?.plusSeconds(LIFETIME_SECONDS)

    /** True once the code the user is showing can no longer be validated. */
    fun hasExpired(at: Instant): Boolean {
        val expiry = expiresAt ?: return false
        return !at.isBefore(expiry)
    }

    // MARK: - Lifecycle

    /** Fetches immediately, then keeps refreshing until [stop]. */
    fun start(auth: AuthManager) {
        stop()
        refreshJob = scope.launch {
            while (isActive) {
                refresh(auth)
                delay(REFRESH_INTERVAL_SECONDS.seconds)
            }
        }
    }

    fun stop() {
        refreshJob?.cancel()
        refreshJob = null
    }

    suspend fun refresh(auth: AuthManager) {
        isRefreshing = true
        try {
            // Goes through performActionWithFreshTokens, so an access token
            // older than five minutes is renewed before this call rather than
            // 401ing.
            val response = auth.authorizedGet(ENDPOINT)
            if (response.status != 200) {
                errorMessage = message(response.status, response.body)
                return
            }
            payload = PREFIX + json.decodeFromString(CodeResponse.serializer(), response.body).code
            issuedAt = Instant.now()
            errorMessage = null
        } catch (error: Exception) {
            // The previous code is deliberately kept: it may still be inside its
            // 300s window, and a card that blanks out on a flaky connection is
            // worse than one showing a code that might still scan.
            errorMessage = error.message ?: error.javaClass.simpleName
        } finally {
            isRefreshing = false
        }
    }

    private fun message(status: Int, body: String): String {
        val detail = body.take(120)
        return when (status) {
            401 -> context.getString(R.string.error_code_401, detail)
            403 -> context.getString(R.string.error_code_403, detail)
            else -> context.getString(R.string.error_code_server, status, detail)
        }
    }

    @Serializable
    private data class CodeResponse(val code: String)

    companion object {
        /** MembershipAPI's `CODE_TTL`. After this the code is dead server-side. */
        const val LIFETIME_SECONDS = 300L

        /** Matches the web card's cadence — comfortably inside [LIFETIME_SECONDS]. */
        const val REFRESH_INTERVAL_SECONDS = 250L

        /** The scanner rejects anything not carrying this prefix. */
        private const val PREFIX = "stsa$"
        private const val ENDPOINT = "https://idms.stsa.tw/membership/api/get_code"

        private val json = Json { ignoreUnknownKeys = true }
    }
}

File: ios/MemberApp/Features/Card/MembershipCodeStore.swift
import Foundation
import Observation

/// Fetches and keeps fresh the code behind the member card's QR.
///
/// The server mints a new code per request and stores it in Redis for 300
/// seconds (`CODE_TTL` in MembershipAPI). The web card re-fetches every 250s to
/// stay inside that window; this does the same, but only while the card is on
/// screen — there is no reason to hold a live code in the user's pocket.
@MainActor
@Observable
final class MembershipCodeStore {
    /// MembershipAPI's `CODE_TTL`. After this the code is dead server-side.
    static let lifetime: TimeInterval = 300
    /// Matches the web card's cadence — comfortably inside `lifetime`.
    static let refreshInterval: TimeInterval = 250

    private(set) var payload: String?
    private(set) var issuedAt: Date?
    private(set) var isRefreshing = false
    private(set) var errorMessage: String?

    private var refreshTask: Task<Void, Never>?

    /// The scanner rejects anything not carrying this prefix.
    private static let prefix = "stsa$"
    private static let endpoint = URL(string: "https://idms.stsa.tw/membership/api/get_code")!

    var expiresAt: Date? {
        issuedAt?.addingTimeInterval(Self.lifetime)
    }

    /// True once the code the user is showing can no longer be validated.
    func hasExpired(at moment: Date) -> Bool {
        guard let expiresAt else { return false }
        return moment >= expiresAt
    }

    // MARK: - Lifecycle

    /// Fetches immediately, then keeps refreshing until `stop()`.
    ///
    /// "Immediately" skips the fetch when a code minted moments ago is still
    /// good — which is what `prefetch(using:)` sets up. Re-minting here would
    /// throw away the head start and leave the member watching a spinner right
    /// after they authenticated.
    func start(using auth: AuthManager) {
        stop()
        refreshTask = Task { [weak self] in
            while !Task.isCancelled {
                if await self?.secondsUntilRefresh ?? 0 <= 0 {
                    await self?.refresh(using: auth)
                }
                let wait = await self?.secondsUntilRefresh ?? Self.refreshInterval
                try? await Task.sleep(for: .seconds(wait))
            }
        }
    }

    /// Mints a code without starting the refresh loop.
    ///
    /// Called alongside the biometric prompt rather than after it. The code is
    /// only ever *drawn* behind a successful unlock, so this leaks nothing to
    /// whoever failed the gate — it just removes a round-trip from the moment it
    /// is most felt. If the gate does fail, the caller drops it with `clear()`.
    func prefetch(using auth: AuthManager) async {
        guard secondsUntilRefresh <= 0 else { return }
        await refresh(using: auth)
    }

    /// Drops a code that was minted but never shown.
    func clear() {
        payload = nil
        issuedAt = nil
        errorMessage = nil
    }

    /// Seconds left before the held code should be replaced; 0 when there is
    /// nothing usable.
    private var secondsUntilRefresh: TimeInterval {
        guard let issuedAt else { return 0 }
        return max(0, Self.refreshInterval - Date().timeIntervalSince(issuedAt))
    }

    func stop() {
        refreshTask?.cancel()
        refreshTask = nil
    }

    func refresh(using auth: AuthManager) async {
        isRefreshing = true
        defer { isRefreshing = false }

        do {
            // Goes through performAction, so an access token older than five
            // minutes is renewed before this call rather than 401ing.
            let request = try await auth.authorizedRequest(for: Self.endpoint)
            let (data, response) = try await URLSession.shared.data(for: request)

            if let http = response as? HTTPURLResponse, http.statusCode != 200 {
                throw CodeError.server(status: http.statusCode,
                                       body: String(data: data, encoding: .utf8) ?? "")
            }

            let decoded = try JSONDecoder().decode(CodeResponse.self, from: data)
            payload = Self.prefix + decoded.code
            issuedAt = Date()
            errorMessage = nil
        } catch {
            // The previous code is deliberately kept: it may still be inside its
            // 300s window, and a card that blanks out on a flaky connection is
            // worse than one showing a code that might still scan.
            errorMessage = error.localizedDescription
        }
    }

    private struct CodeResponse: Decodable {
        let code: String
    }

    enum CodeError: LocalizedError {
        case server(status: Int, body: String)

        var errorDescription: String? {
            switch self {
            case .server(let status, let body) where status == 401:
                "登入已失效,請重新登入。(401 \(body.prefix(120)))"
            case .server(let status, let body) where status == 403:
                "帳號缺少必要的授權範圍。(403 \(body.prefix(120)))"
            case .server(let status, let body):
                "伺服器錯誤 \(status)。\(body.prefix(120))"
            }
        }
    }
}
